core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -4367,10 +4443,49 @@ class KafkaApis(val requestChannel: RequestChannel,
     CompletableFuture.completedFuture[Unit](())
   }
 
-  private def getAcknowledgeBatchesFromShareFetchRequest(shareFetchRequest: ShareFetchRequest,
-                                                         topicIdNames: util.Map[Uuid, String],
-                                                         erroneous: mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]
-                                                        ): mutable.Map[TopicIdPartition, util.List[ShareAcknowledgementBatch]] = {
+  // Visible for Testing
+  def getAcknowledgeBatchesFromShareAcknowledgeRequest(shareAcknowledgeRequest: ShareAcknowledgeRequest,
+                                                       topicIdNames: util.Map[Uuid, String],
+                                                       erroneous: mutable.Map[TopicIdPartition, ShareAcknowledgeResponseData.PartitionData]
+                                                      ): mutable.Map[TopicIdPartition, util.List[ShareAcknowledgementBatch]] = {
+    val acknowledgeBatchesMap = mutable.Map[TopicIdPartition, util.List[ShareAcknowledgementBatch]]()
+    shareAcknowledgeRequest.data().topics().forEach{ topic =>
+      if (!topicIdNames.containsKey(topic.topicId)) {
+        topic.partitions.forEach{ case partition: ShareAcknowledgeRequestData.AcknowledgePartition =>
+          val topicIdPartition = new TopicIdPartition(
+            topic.topicId,
+            new TopicPartition(null, partition.partitionIndex))
+          erroneous +=
+            topicIdPartition -> ShareAcknowledgeResponse.partitionResponse(topicIdPartition, Errors.UNKNOWN_TOPIC_ID)
+        }
+      } else {
+        topic.partitions().forEach{ partition =>
+          if (partition.acknowledgementBatches().size() > 0) {
```

Comment:
As Apoorv mentioned earlier, by excluding partitions with empty ack batches, we won't include that partition in the response, which is incorrect, right?

Reply:
Thanks for the review. That is the incorrect behaviour. I have made changes in handleAcknowledgements method to take care of this, and removed the check `if (partition.acknowledgementBatches().size() > 0)` altogether from `getAcknowledgeBatchesFromShareAcknowledgeRequest` and `getAcknowledgeBatchesFromShareFetchRequest`